<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">

<title>Introduction to Bioinformatics</title>

<meta name="description" content="Introduction to Bioinformatics">

<!-- Disabled for now - embeds HTML in author string -->
<!--  <meta name="author" content="Lesson 1 - Gestalt &amp; remote access" /> -->

<meta name="apple-mobile-web-app-capable" content="yes" />
<meta name="apple-mobile-web-app-status-bar-style" content="black-translucent" />

<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">

<link rel="stylesheet" href="reveal.js/css/reveal.css">
  <link rel="stylesheet" href="src/theme/csmall.css" id="theme">


<!-- For syntax highlighting -->
  <link rel="stylesheet" href="reveal.js/lib/css/zenburn.css">


<!-- If the query includes 'print-pdf', use the PDF print sheet -->
<script>
  document.write( '<link rel="stylesheet" href="reveal.js/css/print/' +
    ( window.location.search.match( /print-pdf/gi ) ? 'pdf' : 'paper' ) +
    '.css" type="text/css" media="print">' );
</script>

<!--[if lt IE 9]>
<script src="reveal.js/lib/js/html5shiv.js"></script>
<![endif]-->

<!-- Note escaped dollars -->
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      tex2jax: {inlineMath: [['$','$'], ['\\(','\\)']]}
    });
  </script>

</head>

<body>

<div class="reveal">

<!-- Any section element inside of this container is displayed as a slide -->
<div class="slides">

<section>
<h1>Introduction to Bioinformatics</h1>
<h3>Lesson 1 - Gestalt &amp; remote access</h3>
<p>
<h4></h4>
</p>
</section>


<section id="first-things-first" class="level1">
<h1>First things first</h1>
<p>Did you</p>
<ul>
<li>read the first couple chapters?</li>
<li>successfully ssh into <code>rhino</code>?</li>
</ul>
</section>
<section id="introductions" class="level1">
<h1>Introductions</h1>
<ul>
<li>Name</li>
<li>What you’re working on</li>
<li>Experience</li>
</ul>
</section>
<section id="what-youll-get-from-this-class" class="level1">
<h1>What you’ll get from this class</h1>
<ul>
<li>High level overview of how to <em>approach</em> bioinformatics</li>
<li>Basic bioinformatics technical skills</li>
<li>Enough momentum to initiate self-learning process</li>
</ul>
</section>
<section id="what-you-wont-get-from-this-class" class="level1">
<h1>What you won’t get from this class</h1>
<p>…but that might be useful:</p>
<ul>
<li>Higher level programming</li>
<li>Statistics</li>
<li>Other mathematics</li>
<li>Knowledge of every bioinformatics program/method ever created</li>
</ul>
<p>But hopefully you’ll have enough momentum to learn these on your own as needed.</p>
</section>
<section id="please-spend-some-time-outside-class" class="level1">
<h1>Please spend <em>some</em> time outside class</h1>
<ul>
<li>Time is the only way to learn these things</li>
<li>The waiting list is very long</li>
</ul>
</section>
<section id="overview-of-class" class="level1">
<h1>Overview of class</h1>
<ol type="1">
<li>Gestalt &amp; remote access</li>
<li>Unix shell - overview/intro</li>
<li>Unix shell - scripting &amp; advanced tools</li>
<li>Version control with git</li>
<li>Python - syntax &amp; types</li>
<li>Python - data structures</li>
<li>Python - writing a program</li>
</ol>
</section>
<section id="today" class="level1">
<h1>Today</h1>
<ul>
<li>1/2 overview</li>
<li>1/2 remote access &amp; Unix environment</li>
</ul>
</section>
<section id="why-computers" class="level1">
<h1>Why computers?</h1>
<p><br/></p>
<p><strong>DATA</strong></p>
</section>
<section id="biology-data" class="level1">
<h1>Biology data</h1>
<p>We can now get it</p>
<ul>
<li>Cheaper &amp; Faster</li>
</ul>
<p>=&gt;</p>
<ul>
<li>We have more of it</li>
</ul>
</section>
<section id="to-understand-biology-we-need-to-understand-data" class="level1">
<h1>To understand biology, we need to understand data</h1>
<p><br/></p>
<p><strong>ENTER BIOINFORMATICS</strong></p>
</section>
<section id="what-is-bioinformatics" class="level1">
<h1>What is bioinformatics?</h1>
<p>The art of applying computation towards the analysis of biological data</p>
</section>
<section id="aspects-of-bioinformatics" class="level1">
<h1>Aspects of Bioinformatics</h1>
<p>Things we typically do with data:</p>
<p>munging | exploration | draw conclusions</p>
</section>
<section id="data-munging" class="level1">
<h1>Data munging</h1>
<ul>
<li>Cleaning data</li>
<li>Transforming data</li>
</ul>
</section>
<section id="data-exploration" class="level1">
<h1>Data exploration</h1>
<ul>
<li>Visualize</li>
<li>Build intuition</li>
<li>Hypothesize</li>
</ul>
</section>
<section id="building-conclusions" class="level1">
<h1>Building conclusions</h1>
<p>In which we shore things up to tell the world…</p>
<ul>
<li>Statistical analysis</li>
<li>More visualization</li>
</ul>
</section>
<section id="rinse-and-repeat" class="level1">
<h1>Rinse and repeat…</h1>
<p>New insights and conclusions lead to new data and questions, etc.</p>
</section>
<section id="so-this-is-what-we-do" class="level1">
<h1>So this is <em>what</em> we do</h1>
<p><br/></p>
<section id="but-how-should-we-do-it" class="level2">
<h2>But <em>how</em> should we do it?</h2>
</section>
</section>
<section id="specific-goals-for-research" class="level1">
<h1>Specific goals for research</h1>
<ul>
<li>Reproducibility</li>
<li>Robustness</li>
<li>Accuracy</li>
<li>Clarity</li>
<li>Iterability</li>
</ul>
</section>
<section id="things-we-can-do-towards-these-goals" class="level1">
<h1>Things we can do towards these goals</h1>
<p>immutability | documentation | automation</p>
</section>
<section id="treat-data-as-immutable" class="level1">
<h1>Treat data as immutable!</h1>
<p><strong>Never</strong> change data in place; data in, data out.</p>
<p>Aids: reproducibility, clarity &amp; iterability</p>
</section>
<section id="document-everything" class="level1">
<h1>Document <strong>everything</strong>!!!</h1>
<p>Hopefully you’re already doing this with your biology work (lab notebook).</p>
<p>Bioinformatics should be no different.</p>
<ul>
<li>Software versions</li>
<li>Settings used to run things</li>
<li>Reasoning behind the way you do things</li>
<li>How/when you got your data</li>
</ul>
<p>Aids: reproducibility, clarity</p>
</section>
<section id="the-best-documentation-is-automation" class="level1">
<h1>The best documentation is automation</h1>
<p>Write <em>scripts</em> (little computer programs) for everything:</p>
<ul>
<li>Munging</li>
<li>Plotting</li>
<li>Statistics</li>
<li>Build scripts (tying it all together)</li>
</ul>
<p>Aids: reproducibility, robustness, accuracy, iterability</p>
</section>
<section id="guiding-principals-practices-towards-this-end" class="level1">
<h1>Guiding principals &amp; practices towards this end</h1>
<p>Unix philosophy | project/data organization</p>
</section>
<section id="the-unix-philosophy" class="level1">
<h1>The Unix Philosophy</h1>
<ul>
<li>Small, composable tools that do one thing right</li>
<li>Plumb tools together into pipelines &amp; scripts</li>
<li>Embrace plain-text data</li>
</ul>
<p>Bioinformatics naturally embraces this.</p>
</section>
<section id="project-data-layout" class="level1">
<h1>Project &amp; data layout</h1>
<p>Being consistent aids in automation &amp; clarity</p>
<ul>
<li>naming schemes</li>
<li>path schemes</li>
</ul>
<p><em>Note: I use a somewhat different organizational scheme than the book’s author. Find what works for you…</em></p>
</section>
<section id="project-layout" class="level1">
<h1>Project layout</h1>
<p>Here’s how I typically organize things:</p>
<pre class="text"><code>your-project
├── README.md
├── build.sh
├── data
│   ├── sequences.fasta
│   └── metadata.csv
├── scripts
│   ├── clean.py
│   └── plot.R
└── output
    ├── alignment.fasta
    ├── cleaned_metadata.csv
    ├── tree.newick
    └── tree_plot.png</code></pre>
</section>
<section id="questions" class="level1">
<h1>Questions?</h1>
</section>
<section id="the-unix-environment" class="level1">
<h1>The Unix environment</h1>
<p>Unix is a proprietary operating system from the 60s.</p>
<p>True Unix isn’t frequently encountered these days, but its philosophy and approach to computing have thrived in “Unix-like” systems. So much so that when we say “Unix”, we usually mean it in the abstract sense.</p>
<p>We’ll be using Ubuntu Linux, a Unix-like operating system.</p>
</section>
<section id="the-unix-shell" class="level1">
<h1>The Unix shell</h1>
<p>One of the core aspects of the Unix environment is the <em>Unix shell</em>.</p>
<p>The shell is just a wrapper around the operating system, through which we can use text commands and output to interact with our computer.</p>
<p>This is analogous to the desktop environment on your personal computer, which has a graphical rather than a textual user interface.</p>
</section>
<section id="remote-servers" class="level1">
<h1>Remote servers</h1>
<p>Unix was built for remote access, since computers were big and expensive.</p>
</section>
<section id="why-we-care-about-remote-servers" class="level1">
<h1>Why we care about remote servers</h1>
<ul>
<li>The Hutch servers are really powerful</li>
<li>We can leave long running computations running without worrying about our computer shutting down by necessity or accident</li>
<li>We can access running programs and data from any computer capable of connecting to the remote server</li>
<li>We can distribute tasks across compute clusters, such as the one the hutch has (See <a href="http://www.fredhutch.io/events/2015/05/20/spring-2015-unix-hpc/">gizmo class</a>)</li>
</ul>
</section>
<section id="ssh-for-remote-connection" class="level1">
<h1>SSH for remote connection</h1>
<p>Do this however you figured out how to do it for your OS.</p>
<pre><code>ssh &lt;username&gt;@rhino</code></pre>
<p>Write down which rhino you connect to, so you can directly connect to that next time.</p>
<p>(e.g If you get connected to <code>rhino03</code>, next time use <code>ssh &lt;username&gt;@rhino03</code>.)</p>
</section>
<section id="youre-now-in-a-unix-shell" class="level1">
<h1>You’re now in a Unix shell!</h1>
<p>We’ll cover things more thoroughly next class. For now just some very basic information for orientation.</p>
</section>
<section id="brief-overview-of-a-unix-command" class="level1">
<h1>Brief overview of a Unix command</h1>
<p><code>command [flags] [operands]</code></p>
<p>Example: <code>ls -a ~</code></p>
<ul>
<li><code>ls</code>: The command, for listing directory contents</li>
<li><code>-a</code>: A flag which specifies that hidden files should be listed</li>
<li><code>~</code>: An operand, in this case a special symbol which points to your home directory</li>
</ul>
</section>
<section id="getting-help-for-a-unix-command" class="level1">
<h1>Getting help for a Unix command</h1>
<p>Most Unix programs will take either a <code>-h</code>, <code>--help</code>, or <code>-help</code> flag and return useful information about how to construct a valid command.</p>
<p>Some commands also have <em>man pages</em>, which you can access using <code>man &lt;command-name&gt;</code>.</p>
<p>Example: <code>ls --help</code></p>
</section>
<section id="lets-get-some-more-practice-with-simple-commands" class="level1">
<h1>Let’s get some more practice with simple commands</h1>
</section>
<section id="create-a-project-directory" class="level1">
<h1>Create a project directory</h1>
<p>This class will build around analysis of a single data set, so let’s set up a project directory for it.</p>
<pre><code>mkdir ~/bioinfclass

cd ~/bioinfclass
pwd

mkdir data output scripts

touch README.md</code></pre>
</section>
<section id="download-data-with-wget" class="level1">
<h1>Download data with <code>wget</code></h1>
<pre><code>wget https://raw.githubusercontent.com/fredhutchio/intro-bioinformatics/gh-pages/data.tar</code></pre>
<p>Note that we get a <code>tar</code> file. This is an archive type, like <code>zip</code>, common on Unix systems.</p>
</section>
<section id="unpacking-the-tar" class="level1">
<h1>Unpacking the tar</h1>
<pre><code>tar -vxf data.tar</code></pre>
</section>
<section id="lets-look-at-what-weve-done" class="level1">
<h1>Let’s look at what we’ve done</h1>
<p><code>tree</code> prints out a directory tree as ASCII art.</p>
<pre><code>tree</code></pre>
</section>
<section id="lets-explore-the-sequence-file" class="level1">
<h1>Let’s explore the sequence file</h1>
<p><code>cat</code> prints the output of file[s] to the screen.</p>
<pre><code>cat data/sfv.fasta</code></pre>
</section>
<section id="now-with-less" class="level1">
<h1>Now with <code>less</code></h1>
<p><code>less</code> lets us “page” through data, without flooding the screen.</p>
<pre><code>less data/sfv.fasta</code></pre>
<p><br/></p>
<p>Press the <code>q</code> key to exit from <code>less</code>.</p>
</section>
<section id="now-lets-look-at-the-metadata" class="level1">
<h1>Now let’s look at the metadata</h1>
<pre><code>less data/sfv.csv</code></pre>
</section>
<section id="supercharging-our-shell" class="level1">
<h1>Supercharging our shell</h1>
<p>SciComp has set up a <em>module</em> system for setting up custom environments.</p>
<p>We’ll be using the <code>intro-bio</code> module, loaded by executing <code>module load intro-bio</code> (note: this <em>only affects the current shell session!</em>). When you do this, you should see something like:</p>
<pre><code>matsen@rhino01:~$ module load intro-bio
Using already loaded python
matsen@rhino01:~$</code></pre>
<p>If you get an error message instead, let us know, because you can’t proceed without this module loaded.</p>
</section>
<section id="tmux---terminal-multiplexer" class="level1">
<h1>Tmux - Terminal multiplexer</h1>
<p>This is a tool that let’s us have multiple Unix shells running at once.</p>
<p>We could just open multiple terminal windows, each with an ssh connection into the remote server. But with tmux, if we close or lose our connection to the server, our terminal session doesn’t get killed; running programs keep running.</p>
</section>
<section id="setting-up-tmux" class="level1">
<h1>Setting up tmux</h1>
<p>We’ll use tmux throughout this class to keep sessions running. But first let’s download a nice tmux configuration file for making things easier.</p>
<pre><code># This is a command that download
wget https://raw.githubusercontent.com/fredhutchio/intro-bioinformatics/gh-pages/config/tmux.conf

ls

# tmux knows to look for our config file at ~/.tmux.conf
mv tmux.conf ~/.tmux.conf

# Note that this file is hidden once we make it a &quot;dotfile&quot;
ls ~
ls -a ~</code></pre>
</section>
<section id="starting-a-new-tmux-session" class="level1">
<h1>Starting a new tmux session</h1>
<pre><code>tmux</code></pre>
<p>You should now see a fresh shell session inside of tmux.</p>
</section>
<section id="quick-tmux-cheatsheet" class="level1">
<h1>Quick tmux cheatsheet</h1>
<ul>
<li><code>Ctrl-a</code> - your “Command key”</li>
<li><code>Ctrl-a ?</code> - Help</li>
<li><code>Ctrl-a d</code> - Detatch session</li>
</ul>
<p><br/></p>
<ul>
<li><code>Ctrl-a c</code> - New window</li>
<li><code>Ctrl-a &lt;Space&gt;</code> - Next window</li>
</ul>
</section>
<section id="looking-at-tmux-in-action" class="level1">
<h1>Looking at tmux in action</h1>
<p>Why we might want to use tmux:</p>
<ul>
<li>Keep working on other things while a long running computation runs</li>
<li>Switch back and forth quickly between multiple projects</li>
</ul>
<p>Let’s see an example.</p>
</section>
<section id="making-a-sequence-alignment" class="level1">
<h1>Making a sequence alignment</h1>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="co"># First start the long running alignment</span>
<span class="kw">muscle</span> -maxiters 2 -in data/sfv.fasta -out output/alignment.fasta

<span class="co"># Get bored waiting and want to work on something else...</span>
<span class="kw">&lt;Ctrl-a</span> c<span class="kw">&gt;</span> <span class="co"># create new window</span>
<span class="kw">cd</span> ~
<span class="kw">mkdir</span> new-project
<span class="kw">cd</span> new-project
<span class="kw">mkdir</span> data output scripts

<span class="co"># Let&#39;s go check on our alignment</span>
<span class="kw">&lt;Ctrl-a</span> <span class="kw">&lt;</span>Space<span class="kw">&gt;&gt;</span>

<span class="co"># Still running... switch back; etc.</span></code></pre>
</section>
<section id="attaching-and-detaching-a-tmux-session" class="level1">
<h1>Attaching and detaching a tmux session</h1>
<p>SSH sessions can disconnect because of:</p>
<ul>
<li>a lost network connection</li>
<li>a computer shutdown (power outage; frozen screen; system updates; etc.)</li>
</ul>
<p>We might also want to stop work on one machine and pick it up on another.</p>
</section>
<section id="detaching-and-attaching-a-tmux-session" class="level1">
<h1>Detaching and attaching a tmux session</h1>
<ol type="1">
<li>Close your terminal window, and ignore any warning prompts.</li>
<li>Reconnect just the way you did at the beginning of the class.</li>
<li>Type <code>tmux attach</code> to attach.</li>
</ol>
</section>
<section id="manual-detach" class="level1">
<h1>Manual detach</h1>
<p>While in tmux, we can explicitly detach using <code>&lt;Ctrl-a d&gt;</code> or by typing <code>tmux detatch</code>.</p>
<p>Try doing this and then reattaching.</p>
</section>
<section id="checking-on-our-alignment" class="level1">
<h1>Checking on our alignment</h1>
<p>Yay! It’s done!</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">less</span> output/alignment.fasta</code></pre>
</section>
<section id="questions-1" class="level1">
<h1>Questions?</h1>
</section>
<section id="homework" class="level1">
<h1>Homework</h1>
<p>Recommended reading:</p>
<ul>
<li>Chapters 1 &amp; 2 if you haven’t already.</li>
<li>Chapters 2 &amp; 3 for basic Unix stuff &amp; tmux.</li>
</ul>
<p>Reading for next class:</p>
<ul>
<li>Chapter 3, and chapter 7 (till around page 148).</li>
</ul>
</section>
<section id="resources" class="level1">
<h1>Resources</h1>
<ul>
<li>These slides: <a href="http://fredhutchio.github.io/intro-bioinformatics">http://fredhutchio.github.io/intro-bioinformatics</a></li>
<li>Unix command reference: <a href="https://ubuntudanmark.dk/filer/fwunixref.pdf">https://ubuntudanmark.dk/filer/fwunixref.pdf</a></li>
<li>Tmux tutorial: <a href="http://www.fredhutch.io/articles/2014/04/27/terminal-multiplex">http://www.fredhutch.io/articles/2014/04/27/terminal-multiplex</a></li>
</ul>
</section>
</div>

<script src="reveal.js/lib/js/head.min.js"></script>
<script src="reveal.js/js/reveal.min.js"></script>

<script>
  // Full list of configuration options available here:
  // https://github.com/hakimel/reveal.js#configuration
  Reveal.initialize({
    controls: true,
    progress: true,
    history: true,
    center: false,


    rollingLinks: false,
  
  // available themes are in /css/theme
      theme: Reveal.getQueryHash().theme || 'csmall',
  
  // default/cube/page/concave/zoom/linear/fade/none
      transition: Reveal.getQueryHash().transition || 'fade',
  
  // Optional libraries used to extend on reveal.js
  dependencies: [
    { src: 'reveal.js/lib/js/classList.js', condition: function() { return !document.body.classList; } },
    { src: 'reveal.js/plugin/markdown/showdown.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
    { src: 'reveal.js/plugin/markdown/markdown.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
    { src: 'reveal.js/plugin/highlight/highlight.js', async: true, callback: function() { hljs.initHighlightingOnLoad(); } },
    { src: 'reveal.js/plugin/zoom-js/zoom.js', async: true, condition: function() { return !!document.body.classList; } },
    { src: 'MathJax/MathJax.js?config=TeX-AMS-MML_HTMLorMML'}
    // { src: 'reveal.js/plugin/notes/notes.js', async: true, condition: function() { return !!document.body.classList; } }
    // { src: 'reveal.js/plugin/remotes/remotes.js', async: true, condition: function() { return !!document.body.classList; } }
  ]
  });
Reveal.addEventListener( 'slidechanged', function( event ) {
    MathJax.Hub.Rerender();
} );
</script>

</body>
</html>
