<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">

<title>Introduction to Bioinformatics</title>

<meta name="description" content="Introduction to Bioinformatics">

<!-- Disabled for now - embeds HTML in author string -->
<!--  <meta name="author" content="Lesson 1 - Gestalt &amp; remote access" /> -->

<meta name="apple-mobile-web-app-capable" content="yes" />
<meta name="apple-mobile-web-app-status-bar-style" content="black-translucent" />

<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">

<link rel="stylesheet" href="reveal.js/css/reveal.css">
  <link rel="stylesheet" href="src/theme/csmall.css" id="theme">


<!-- For syntax highlighting -->
  <link rel="stylesheet" href="reveal.js/lib/css/zenburn.css">


<!-- If the query includes 'print-pdf', use the PDF print sheet -->
<script>
  document.write( '<link rel="stylesheet" href="reveal.js/css/print/' +
    ( window.location.search.match( /print-pdf/gi ) ? 'pdf' : 'paper' ) +
    '.css" type="text/css" media="print">' );
</script>

<!--[if lt IE 9]>
<script src="reveal.js/lib/js/html5shiv.js"></script>
<![endif]-->

<!-- Note escaped dollars -->
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      tex2jax: {inlineMath: [['$','$'], ['\\(','\\)']]}
    });
  </script>

</head>

<body>

<div class="reveal">

<!-- Any section element inside of this container is displayed as a slide -->
<div class="slides">

<section>
<h1>Introduction to Bioinformatics</h1>
<h3>Lesson 1 - Gestalt &amp; remote access</h3>
<p>
<h4></h4>
</p>
</section>


<section id="first-things-first" class="level1">
<h1>First things first</h1>
<p>Did everyone successfully ssh into <code>rhino</code>?</p>
</section>
<section id="introductions" class="level1">
<h1>Introductions</h1>
<ul>
<li>Name</li>
<li>What you’re working on</li>
<li>Technical Experience</li>
</ul>
</section>
<section id="what-youll-get-from-this-class" class="level1">
<h1>What you’ll get from this class</h1>
<ul>
<li>High level overview of how to <em>approach</em> bioinformatics</li>
<li>Basic bioinformatics technical and problem solving skills</li>
<li>Enough momentum to initiate self-learning process</li>
</ul>
</section>
<section id="what-you-wont-get-from-this-class" class="level1">
<h1>What you won’t get from this class</h1>
<p>…but that might be useful:</p>
<ul>
<li>Higher level programming</li>
<li>Statistics</li>
<li>Other mathematics</li>
<li>Knowledge of every bioinformatics program/method ever created</li>
</ul>
<p>But hopefully you’ll have enough momentum to learn these on your own as needed.</p>
</section>
<section id="please-spend-some-time-outside-class" class="level1">
<h1>Please spend <em>some</em> time outside class</h1>
<ul>
<li>Time is the only way to learn these things</li>
<li>Help keep this class moving smoothly</li>
<li>The waiting list is very long</li>
</ul>
</section>
<section id="what-is-bioinformatics" class="level1">
<h1>What is bioinformatics?</h1>
</section>
<section id="what-is-bioinformatics-1" class="level1">
<h1>What is bioinformatics?</h1>
<p>The application of computation towards the analysis of biological data</p>
</section>
<section id="in-practice-this-entails" class="level1">
<h1>In practice this entails</h1>
<p>data munging | data exploration | drawing conclusions</p>
</section>
<section id="data-munging" class="level1">
<h1>Data munging</h1>
<p>This is where a lot of day to day work goes</p>
<ul>
<li>Cleaning data</li>
<li>Transforming data</li>
</ul>
</section>
<section id="data-exploration" class="level1">
<h1>Data exploration</h1>
<ul>
<li>Visualize</li>
<li>Build intuition</li>
<li>Hypothesize</li>
</ul>
</section>
<section id="building-conclusions" class="level1">
<h1>Building conclusions</h1>
<p>In which we shore things up to tell the world…</p>
<ul>
<li>Statistical analysis</li>
<li>More visualization</li>
</ul>
</section>
<section id="rinse-and-repeat" class="level1">
<h1>Rinse and repeat…</h1>
<p>New insights and conclusions lead to new data and questions, etc.</p>
</section>
<section id="this-should-be-nothing-new" class="level1">
<h1>This should be nothing new</h1>
<p>This is not that different a process from how we typically do science!</p>
<p>The main difference is that here we’re using computers to carry these steps out that we would normally do at the bench or with a piece of paper.</p>
</section>
<section id="so-this-is-what-we-do" class="level1">
<h1>So this is <em>what</em> we do</h1>
<p><br/></p>
<section id="but-how-should-we-do-it" class="level2">
<h2>But <em>how</em> should we do it?</h2>
</section>
</section>
<section id="specific-goals-for-research" class="level1">
<h1>Specific goals for research</h1>
<p>General goals for good science:</p>
<ul>
<li>Reproducibility</li>
<li>Robustness</li>
<li>Accuracy</li>
<li>Clarity</li>
<li>Iterability</li>
</ul>
</section>
<section id="things-we-can-do-towards-these-goals" class="level1">
<h1>Things we can do towards these goals</h1>
<p>Immutable data | Documentation | Automation</p>
</section>
<section id="immutability" class="level1">
<h1>Immutability</h1>
<p><strong>Never</strong> change data in place; data in, data out.</p>
<p>Aids: reproducibility, clarity &amp; iterability</p>
</section>
<section id="document-everything" class="level1">
<h1>Document <strong>everything</strong></h1>
<p>Hopefully you’re already doing this with your biology work (lab notebook).</p>
<p>Bioinformatics should be no different.</p>
<ul>
<li>Software versions</li>
<li>Settings used to run things</li>
<li>Reasoning behind the way you do things</li>
<li>How/when you got your data</li>
</ul>
<p>Aids: reproducibility, clarity</p>
</section>
<section id="the-best-documentation-is-automation" class="level1">
<h1>The best documentation is automation</h1>
<p>Write <em>scripts</em> (little computer programs) for everything:</p>
<ul>
<li>Munging</li>
<li>Plotting</li>
<li>Statistics</li>
<li>Build scripts (tying it all together)</li>
</ul>
<p>Aids: reproducibility, robustness, accuracy, iterability</p>
</section>
<section id="guiding-principles-practices-towards-this-end" class="level1">
<h1>Guiding principles &amp; practices towards this end</h1>
<p>Unix philosophy | project/data organization</p>
</section>
<section id="the-unix-philosophy" class="level1">
<h1>The Unix Philosophy</h1>
<ul>
<li>Small, <strong>composable</strong> tools that do one thing right</li>
<li>Plumb tools together into pipelines &amp; scripts</li>
<li>Embrace plain-text data</li>
</ul>
<p>Bioinformatics naturally embraces this.</p>
</section>
<section id="project-data-layout" class="level1">
<h1>Project &amp; data layout</h1>
<p>Being <strong>consistent</strong> aids in automation &amp; clarity</p>
<ul>
<li>naming schemes</li>
<li>path schemes</li>
</ul>
</section>
<section id="questions" class="level1">
<h1>Questions?</h1>
</section>
<section id="a-bit-on-problem-solving" class="level1">
<h1>A bit on problem solving</h1>
<p>Programmers spend more time sitting around and thinking than they actually do typing code. What follows are some tips along these lines.</p>
</section>
<section id="polyas-how-to-solve-it" class="level1">
<h1>Polya’s How to Solve It</h1>
<p><a href="https://en.wikipedia.org/wiki/How_to_Solve_It" class="uri">https://en.wikipedia.org/wiki/How_to_Solve_It</a></p>
<ol type="1">
<li>Understand the problem</li>
<li>Make a plan</li>
<li>Carry out the plan</li>
<li>Look back and reflect</li>
</ol>
<p>When this fails, try to solve an easier or related problem first.</p>
</section>
<section id="understanding" class="level1">
<h1>Understanding</h1>
<p>This is sometimes the trickiest part, especially when working on a subproblem, because things can be ill defined. I find that spending time clarifying and truly understanding a difficult problem can often gets me 70% of the way to a solution.</p>
<p>Look for the things that feel fuzzy and unclear and try to state precisely what it is you don’t get about them.</p>
</section>
<section id="devising-a-plan" class="level1">
<h1>Devising a plan</h1>
<p>This is most of the rest of the work; The better your plan, the easier it will be to execute. Sometimes perfecting a plan will elighten your understanding as well.</p>
<p>This should be about 20%.</p>
</section>
<section id="execution-and-reflection" class="level1">
<h1>Execution and reflection</h1>
<p>Ideally, just the last 10%, if understanding and planning are perfect (reality can vary).</p>
</section>
<section id="strategies-at-all-stages" class="level1">
<h1>Strategies at all stages:</h1>
<ul>
<li>Draw a picture; visualize</li>
<li>Look for a pattern</li>
<li>Use a model or simpler problem</li>
<li>Work backwards</li>
<li>Eliminate possibilities</li>
<li>List knowns</li>
<li>Run examples</li>
</ul>
</section>
<section id="more-on-unix" class="level1">
<h1>More on Unix</h1>
<p>Unix is a proprietary operating system from the 60s.</p>
<p>True Unix isn’t frequently encountered these days, but its philosophy and approach to computing have thrived in “Unix-like” systems. So much so that when we say “Unix”, we usually mean it in the abstract sense.</p>
<p>We’ll be using Ubuntu Linux, a Unix-like operating system.</p>
</section>
<section id="the-unix-shell" class="level1">
<h1>The Unix shell</h1>
<p>One of the core aspects of the Unix environment is the <em>Unix shell</em>.</p>
<p>The shell is just a wrapper around the operating system, through which we can use text commands and output to interact with our computer.</p>
<p>This is analogous to the desktop environment on your personal computer, which has a graphical rather than a textual user interface.</p>
</section>
<section id="remote-servers" class="level1">
<h1>Remote servers</h1>
<p>Unix was built for remote access, since computers were big and expensive.</p>
</section>
<section id="why-we-care-about-remote-servers" class="level1">
<h1>Why we care about remote servers</h1>
<ul>
<li>The Hutch servers are really powerful</li>
<li>We can leave long running computations running without worrying about our computer shutting down by necessity or accident</li>
<li>We can access running programs and data from any computer capable of connecting to the remote server</li>
<li>We can distribute tasks across compute clusters, such as the one the Hutch has (See <a href="http://www.fredhutch.io/events/2015/05/20/spring-2015-unix-hpc/">gizmo class</a>)</li>
</ul>
</section>
<section id="questions-1" class="level1">
<h1>Questions?</h1>
</section>
<section id="practicum" class="level1">
<h1>Practicum!</h1>
<ul>
<li>Connect to a remote server</li>
<li>Unix survival guide</li>
<li>Set up a project directory and load it with data</li>
<li>Investigate our data</li>
<li>Learn how to edit files on a server</li>
<li>Tmux for session management</li>
</ul>
</section>
<section id="ssh-for-remote-connection" class="level1">
<h1>SSH for remote connection</h1>
<p>Do this however you figured out how to do it for your OS.</p>
<pre><code>ssh &lt;username&gt;@rhino</code></pre>
<p>Write down which rhino you connect to, so you can directly connect to that next time. (e.g If you get connected to <code>rhino3</code>, next time use <code>ssh &lt;username&gt;@rhino3</code>.)</p>
<p>Let us know if you get an error message along the lines of “<code>can't find home</code>”.</p>
</section>
<section id="from-outside-the-hutch" class="level1">
<h1>From outside the Hutch</h1>
<p>Side note:</p>
<p>If you want to work on the servers from outside the Hutch, you can ssh into <code>&lt;username&gt;@snail.fhcrc.org</code>, and then from there execute <code>ssh &lt;username&gt;@rhino&lt;N&gt;</code> to get into your rino.</p>
</section>
<section id="youre-now-in-a-unix-shell" class="level1">
<h1>You’re now in a Unix shell!</h1>
<p>We’ll cover things more thoroughly next class. For now, just some basic information for orientation.</p>
<p>What follows is your Zombie Apocalypse Unix guide.</p>
</section>
<section id="brief-overview-of-a-unix-command" class="level1">
<h1>Brief overview of a Unix command</h1>
<p><code>command [flags] [operands]</code></p>
<pre><code>ls -a ~</code></pre>
<ul>
<li><code>ls</code>: The command, in this case for listing directory contents</li>
<li><code>-a</code>: A flag which specifies that hidden files should be listed</li>
<li><code>~</code>: An operand, in this case a special symbol which points to your home directory</li>
</ul>
<p>Note: Spaces are important here (but not how many).</p>
</section>
<section id="getting-help-for-a-unix-command" class="level1">
<h1>Getting help for a Unix command</h1>
<p>Most Unix programs will take either a <code>-h</code>, <code>--help</code>, or <code>-help</code> flag and return useful information about how to construct a valid command.</p>
<p>Some commands also have <em>man pages</em>, which you can access using <code>man &lt;command-name&gt;</code>.</p>
<p>Example: <code>ls --help</code> &amp; <code>man ls</code></p>
</section>
<section id="stuck-in-a-terminal" class="level1">
<h1>Stuck in a terminal?</h1>
<p>If you are in the terminal and things seem “stuck” (a program is running that won’t stop), try the following:</p>
<p><code>Ctrl-c</code>, <code>Esc</code>, <code>q</code>, <code>Ctrl-d</code>, <code>Esc : q Enter</code></p>
<p>Note: These are <em>key commands</em>;</p>
<ul>
<li><code>-</code> means press the keys at the same time.</li>
<li>` ` means let go and press the next key</li>
</ul>
</section>
<section id="autocomplete" class="level1">
<h1>Autocomplete</h1>
<p>Bash can “auto-complete” command and filenames. Just start typing and hit <code>Tab</code>.</p>
<p>Try <code>ls ~/bioi&lt;Tab&gt;</code>.</p>
</section>
<section id="history" class="level1">
<h1>History</h1>
<p>Bash maintains a history of the last several commands you executed. You can access these by typing <code>history</code>.</p>
<p>If you want to run a command very close to one you just executed (but perhaps with some parameters tweaked), you can use the up and down arrow keys to move through the history, edit the line, and re-execute.</p>
</section>
<section id="questions-2" class="level1">
<h1>Questions?</h1>
</section>
<section id="setting-up-our-project" class="level1">
<h1>Setting up our project</h1>
<p>This class will build around analysis of a <a href="http://dx.doi.org/10.1038/emi.2013.60">real world dataset</a> of <a href="https://en.wikipedia.org/wiki/Simian_foamy_virus">Simian Foamy Virus</a> (SFV). SFV is a retrovirus that typically infects non-human primates (NHP), but can infect humans bitten by NHP (though we’re a “dead-end” host). This data set looks at viruses sampled from humans and monkeys in Bangladesh.</p>
</section>
<section id="project-layout" class="level1">
<h1>Project layout</h1>
<p>Here’s how I typically organize things:</p>
<pre class="text"><code>your-project
├── README.md
├── build.sh
├── data
│   ├── sequences.fasta
│   └── metadata.csv
├── scripts
│   ├── clean.py
│   └── plot.R
└── output
    ├── alignment.fasta
    ├── cleaned_metadata.csv
    ├── tree.newick
    └── tree_plot.png</code></pre>
<p>Note: this is a little different than the organization from the book; Use what works for you, but be consistent.</p>
</section>
<section id="setting-up-our-project-directory" class="level1">
<h1>Setting up our project directory</h1>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">mkdir</span> ~/bioinfclass

<span class="kw">cd</span> ~/bioinfclass
<span class="kw">mkdir</span> data output scripts

<span class="kw">ls</span></code></pre></div>
<p>(You’ll always enter in the things you see in these black boxes)</p>
</section>
<section id="download-data-with-wget" class="level1">
<h1>Download data with <code>wget</code></h1>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">wget</span> https://goo.gl/8Nk5tZ -O data.tar
<span class="kw">ls</span></code></pre></div>
<p>Note that we now have a <code>data.tar</code> file in this directory. This is an archive type (like <code>zip</code>) common on Unix systems. To unpack it</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">tar</span> -v -x -f data.tar</code></pre></div>
</section>
<section id="editing-text-files-with-nano" class="level1">
<h1>Editing text files with <code>nano</code></h1>
<p>Type <code>nano README.md</code> to open a new text file called <code>README.md</code></p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">nano</span> README.md</code></pre></div>
<p>You are now in the <code>nano</code> text editing program. Type your thoughts about the class so far…</p>
<p>To save, type <code>Ctrl-o</code> and hit enter. To exit type <code>Ctrl-x</code> (if there are unsaved changes, next type <code>y</code> then <code>Enter</code> to save them).</p>
</section>
<section id="lets-look-at-what-weve-done" class="level1">
<h1>Let’s look at what we’ve done</h1>
<p><code>tree</code> prints out a directory tree as ASCII art.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">
<span class="kw">tree</span></code></pre></div>
<p>Note: This may not turn out right on Windows PuTTY terminals.</p>
</section>
<section id="lets-explore-the-sequence-file" class="level1">
<h1>Let’s explore the sequence file</h1>
<p><code>cat</code> prints the output of file(s) to the screen.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">cat</span> data/sfv.fasta</code></pre></div>
</section>
<section id="now-with-less" class="level1">
<h1>Now with <code>less</code></h1>
<p><code>less</code> lets us “page” through data, without flooding the screen.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">less</span> data/sfv.fasta</code></pre></div>
<p><br/></p>
<p>Press the <code>q</code> key to exit from <code>less</code>.</p>
</section>
<section id="now-lets-look-at-the-metadata" class="level1">
<h1>Now let’s look at the metadata</h1>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">less</span> data/sfv.csv</code></pre></div>
<p>Press the <code>/</code> key to enter a search string</p>
<p>(Try <code>human</code> or <code>monkey</code>; Use <code>n</code> and <code>N</code> to toggle through results).</p>
</section>
<section id="questions-3" class="level1">
<h1>Questions?</h1>
</section>
<section id="supercharging-our-shell" class="level1">
<h1>Supercharging our shell</h1>
<p>SciComp has set up a <em>module</em> system for customizing your environment.</p>
<p>We’ll be using the <code>intro-bio</code> module, loaded by executing <code>module load intro-bio</code> (note: this <em>only affects the current shell session!</em>). When you do this, you should see something like:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">module</span> load intro-bio</code></pre></div>
<p>You might get a message that says <code>Using already loaded python</code>, but don’t worry about it. If you get an error message though, let us know. You can’t proceed without this module loaded.</p>
</section>
<section id="module-system-side-note" class="level1">
<h1>Module system side note</h1>
<p><strong>Note:</strong> if you <em>ever</em> get a “command not found” error, you probably need to re-load your modules.</p>
</section>
<section id="tmux---terminal-multiplexer" class="level1">
<h1>Tmux - Terminal multiplexer</h1>
<p>This is a tool that let’s us have multiple Unix shells running at once.</p>
<p>We could just open multiple terminal windows, each with an ssh connection into the remote server. But with tmux, if we close or lose our connection to the server, our terminal session doesn’t get killed; running programs keep running. This lets us:</p>
<ul>
<li>Put down and pick up work from multiple machines</li>
<li>Keep working on other things while a long running computation runs</li>
<li>Switch back and forth quickly between multiple projects</li>
</ul>
</section>
<section id="setting-up-tmux" class="level1">
<h1>Setting up tmux</h1>
<p>We’ll use tmux throughout this class to keep sessions running. But first let’s download a nice tmux configuration file for making things easier.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co"># This is a command that downloads a tmux configuration file</span>
<span class="kw">wget</span> https://raw.githubusercontent.com/fredhutchio/intro-bioinformatics/gh-pages/config/tmux.conf

<span class="kw">ls</span>

<span class="co"># tmux knows to look for our config file at ~/.tmux.conf</span>
<span class="kw">mv</span> tmux.conf ~/.tmux.conf

<span class="co"># Note that this file is hidden once we make it a &quot;dotfile&quot;</span>
<span class="kw">ls</span> ~
<span class="kw">ls</span> -a ~</code></pre></div>
</section>
<section id="starting-a-new-tmux-session" class="level1">
<h1>Starting a new tmux session</h1>
<pre><code>tmux</code></pre>
<p>You should now see a fresh shell session inside of tmux.</p>
</section>
<section id="quick-tmux-cheatsheet" class="level1">
<h1>Quick tmux cheatsheet</h1>
<ul>
<li><code>Ctrl-a</code> - your “Command key”</li>
<li><code>Ctrl-a ?</code> - Help</li>
<li><code>Ctrl-a d</code> - Detach session</li>
</ul>
<p>Windows:</p>
<ul>
<li><code>Ctrl-a c</code> - New window</li>
<li><code>Ctrl-a &lt;Space&gt;</code> - Next window</li>
</ul>
</section>
<section id="quick-tmux-cheatsheet-1" class="level1">
<h1>Quick tmux cheatsheet</h1>
<p>Panes:</p>
<ul>
<li><code>Ctrl-a |</code> - Split window into panes vertically</li>
<li><code>Ctrl-a -</code> - Split window into panes horizontally</li>
<li><code>Ctrl-a &lt;arrow&gt;</code> - Move between</li>
<li><code>Ctrl-d</code> - Close a pane or window</li>
</ul>
</section>
<section id="tmux-commands" class="level1">
<h1>Tmux commands</h1>
<p>You can also interact with tmux using commands, and you can see a list of commands and explanations on the man page:</p>
<pre><code>man tmux</code></pre>
<p>All of the key combos above are just bindings to these commands, so this is a good place to go if you’re trying to figure something out.</p>
<p>You can also type <code>tmux</code> and hit <code>&lt;Tab&gt;</code> to see a list of commands (this will autocomplete if you type the first part of the command; try <code>tmux kill-</code> then <code>&lt;Tab&gt;</code>).</p>
</section>
<section id="lets-put-tmux-to-the-test" class="level1">
<h1>Lets put Tmux to the test</h1>
<p>And also do our first sequence alignment!</p>
</section>
<section id="what-is-a-sequence-alignment" class="level1">
<h1>What is a sequence alignment?</h1>
<p>As genetic sequences evolve, base pairs can be added or dropped. Alignment tries to figure out where these insertion/deletion events happened so that individual sites in an alignment can be compared between sequences:</p>
<pre><code>AAGGCCTT            AAGGCCTT
ACGGCCTT    ==&gt;&gt;    ACGGCCTT
AAGCCTA             AAG-CCTA
ACGGCCTA            ACGGCCTA</code></pre>
<p>Note: the <code>-</code> character in the alignment represents a putative deletion.</p>
</section>
<section id="making-a-sequence-alignment" class="level1">
<h1>Making a sequence alignment</h1>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co"># First start the long running alignment</span>
<span class="kw">cd</span> ~/bioinfclass
<span class="kw">muscle</span> -maxiters 2 -in data/sfv.fasta -out output/alignment.fasta

<span class="co"># Get bored waiting and want to work on something else...</span>
<span class="kw">&lt;Ctrl-a</span> <span class="kw">|&gt;</span> <span class="co"># create new pane</span>

<span class="kw">cd</span> ~
<span class="kw">mkdir</span> new-project
<span class="kw">cd</span> new-project
<span class="kw">mkdir</span> data output scripts</code></pre></div>
</section>
<section id="detaching-and-attaching-a-tmux-session" class="level1">
<h1>Detaching and attaching a tmux session</h1>
<ol type="1">
<li>Close your terminal window, and ignore any warning prompts.</li>
<li>Reconnect just the way you did at the beginning of the class.</li>
<li>Type <code>tmux attach</code> to attach to an existing session.</li>
</ol>
<p><strong>Remember:</strong> <code>tmux attach</code> will fail if you don’t have a session open already; if that happens just enter <code>tmux</code> to start a new session.</p>
</section>
<section id="manual-detach" class="level1">
<h1>Manual detach</h1>
<p>While in tmux, we can explicitly detach by typing <code>tmux detach</code>.</p>
<p>Try doing this and then reattaching.</p>
</section>
<section id="questions-4" class="level1">
<h1>Questions?</h1>
</section>
<section id="source-code" class="level1">
<h1>Source code!</h1>
<p>The source code for these slides is available on GitHub.</p>
<p>https://github.com/fredhutchio/intro-bioinformatics</p>
<p>For the code that generated these slides, look at the <code>.mds</code> files in the <a href="https://github.com/fredhutchio/intro-bioinformatics/tree/gh-pages/src"><code>src</code></a> directory. (This code is an extension of Markdown, which we’ll look at more next week)</p>
<p>To submit issues or questions about the class, go to the <a href="https://github.com/fredhutchio/intro-bioinformatics/issues">Issues</a> page.</p>
</section>
<section id="exercises" class="level1">
<h1>Exercises</h1>
<p>Please, <em>please</em>, <em><strong>please!</strong></em></p>
<p>Go over <code>tmux</code> and <code>nano</code> and memorize the basics of how to use them. We’ll use these tools over and over and this will help keep things moving.</p>
</section>
<section id="homework" class="level1">
<h1>Homework</h1>
<p>Recommended reading:</p>
<ul>
<li>Chapters 1 &amp; 2 if you haven’t already.</li>
<li>Chapters 2 &amp; 3 for basic Unix stuff &amp; tmux.</li>
</ul>
<p>Reading for next class:</p>
<ul>
<li>Chapter 3, and chapter 7 (till around page 148).</li>
</ul>
</section>
<section id="resources" class="level1">
<h1>Resources</h1>
<ul>
<li>Class home page: <a href="http://fredhutchio.github.io/intro-bioinformatics" class="uri">http://fredhutchio.github.io/intro-bioinformatics</a></li>
<li>Polya: How to solve it <a href="https://en.wikipedia.org/wiki/How_to_Solve_It" class="uri">https://en.wikipedia.org/wiki/How_to_Solve_It</a></li>
<li>Unix command reference: <a href="https://ubuntudanmark.dk/filer/fwunixref.pdf" class="uri">https://ubuntudanmark.dk/filer/fwunixref.pdf</a></li>
<li>Tmux tutorial: <a href="http://www.fredhutch.io/articles/2014/04/27/terminal-multiplex" class="uri">http://www.fredhutch.io/articles/2014/04/27/terminal-multiplex</a></li>
<li>Stack overflow: <a href="https://stackoverflow.com" class="uri">https://stackoverflow.com</a></li>
<li>More on our data: http://dx.doi.org/10.1038/emi.2013.60</li>
</ul>
</section>
</div>

<script src="reveal.js/lib/js/head.min.js"></script>
<script src="reveal.js/js/reveal.min.js"></script>

<script>
  // Full list of configuration options available here:
  // https://github.com/hakimel/reveal.js#configuration
  Reveal.initialize({
    controls: true,
    progress: true,
    history: true,
    center: false,


    rollingLinks: false,
  
  // available themes are in /css/theme
      theme: Reveal.getQueryHash().theme || 'csmall',
  
  // default/cube/page/concave/zoom/linear/fade/none
      transition: Reveal.getQueryHash().transition || 'fade',
  
  // Optional libraries used to extend on reveal.js
  dependencies: [
    { src: 'reveal.js/lib/js/classList.js', condition: function() { return !document.body.classList; } },
    { src: 'reveal.js/plugin/markdown/showdown.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
    { src: 'reveal.js/plugin/markdown/markdown.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
    { src: 'reveal.js/plugin/highlight/highlight.js', async: true, callback: function() { hljs.initHighlightingOnLoad(); } },
    { src: 'reveal.js/plugin/zoom-js/zoom.js', async: true, condition: function() { return !!document.body.classList; } },
    { src: 'MathJax/MathJax.js?config=TeX-AMS-MML_HTMLorMML'}
    // { src: 'reveal.js/plugin/notes/notes.js', async: true, condition: function() { return !!document.body.classList; } }
    // { src: 'reveal.js/plugin/remotes/remotes.js', async: true, condition: function() { return !!document.body.classList; } }
  ]
  });
Reveal.addEventListener( 'slidechanged', function( event ) {
    MathJax.Hub.Rerender();
} );
</script>

</body>
</html>
