<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">

<title>Introduction to Bioinformatics</title>

<meta name="description" content="Introduction to Bioinformatics">

<!-- Disabled for now - embeds HTML in author string -->
<!--  <meta name="author" content="Lesson 3 - Unix Scripting &amp; Automation" /> -->

<meta name="apple-mobile-web-app-capable" content="yes" />
<meta name="apple-mobile-web-app-status-bar-style" content="black-translucent" />

<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">

<link rel="stylesheet" href="reveal.js/css/reveal.css">
  <link rel="stylesheet" href="src/theme/csmall.css" id="theme">


<!-- For syntax highlighting -->
  <link rel="stylesheet" href="reveal.js/lib/css/zenburn.css">


<!-- If the query includes 'print-pdf', use the PDF print sheet -->
<script>
  document.write( '<link rel="stylesheet" href="reveal.js/css/print/' +
    ( window.location.search.match( /print-pdf/gi ) ? 'pdf' : 'paper' ) +
    '.css" type="text/css" media="print">' );
</script>

<!--[if lt IE 9]>
<script src="reveal.js/lib/js/html5shiv.js"></script>
<![endif]-->

<!-- Note escaped dollars -->
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      tex2jax: {inlineMath: [['$','$'], ['\\(','\\)']]}
    });
  </script>

</head>

<body>

<div class="reveal">

<!-- Any section element inside of this container is displayed as a slide -->
<div class="slides">

<section>
<h1>Introduction to Bioinformatics</h1>
<h3>Lesson 3 - Unix Scripting &amp; Automation</h3>
<p>
<h4></h4>
</p>
</section>


<section id="first-some-housekeeping" class="level1">
<h1>First some housekeeping:</h1>
<ul>
<li>Questions from last week?</li>
<li>Try alnvu again: <code>av -L 99999 -w 99999 -c output/alignment.fasta | less -S</code></li>
<li>Tmux bindings?: Replace <code>#{pane_current_path}</code> with <code>$PWD</code> in <code>nano ~/.tmux.conf</code></li>
</ul>
</section>
<section id="shell-scripting" class="level1">
<h1>Shell scripting</h1>
<p><em>In which our flying, speaking beasts assemble and become legion</em></p>
</section>
<section id="the-unix-shell-is-more-than-just-a-shell" class="level1">
<h1>The Unix shell is more than just a shell</h1>
<p>It’s also a mini-programming language!</p>
<p>Building new commands out of old commands is part of Unix composability</p>
</section>
<section id="for-bioinformatics-this-means" class="level1">
<h1>For bioinformatics this means</h1>
<ul>
<li>Simple analysis tools/commands</li>
<li>Automation</li>
</ul>
</section>
<section id="shell-scripting-is-just-duct-tape" class="level1">
<h1>Shell scripting is just “duct tape”</h1>
<ul>
<li>Perfect for connecting existing shell commands together</li>
<li>Less so for more complicated data flow and logic (for which we have python, R, etc.)</li>
</ul>
</section>
<section id="our-first-script" class="level1">
<h1>Our first script!</h1>
<p>We’re going to create a <code>build.sh</code> script to automate our analyses.</p>
</section>
<section id="a-shell-script-is-just-a-list-of-commands-to-be-executed" class="level1">
<h1>A shell script is just a list of commands to be executed</h1>
<p>Run <code>nano build.sh</code>, then add:</p>
<pre><code>ls -l | column -t
echo &quot;IT WORKED!&quot;</code></pre>
<ul>
<li>Save (<code>Ctrl-o</code>)</li>
<li>Switch to a new t-mux pane (<code>Ctrl-a |</code>)</li>
<li>Run <code>bash build.sh</code> from within the <code>bioinfclass</code> directory</li>
</ul>
</section>
<section id="error-handling" class="level1">
<h1>Error handling</h1>
<p>Back in our <code>nano build.sh</code> pane, edit to look like:</p>
<pre><code># Sane error handling settings
set -euf -o pipefail

ls -l | column -t
# Put in a &quot;bug&quot;:
asdfasdfasdf

echo &quot;IT WORKED!&quot;</code></pre>
<p>Save and rerun <code>bash build.sh</code>. Remove the “bug” and rerun again.</p>
<p>Note: Lines starting with <code>#</code> are comments and not executed.</p>
</section>
<section id="making-into-a-proper-command" class="level1">
<h1>Making into a proper command</h1>
<p>Edit your <code>nano build.sh</code> to look like:</p>
<pre><code>#!/bin/bash

# Sane error handling settings
set -euf -o pipefail

ls -l | column -t
echo &quot;IT WORKED!&quot;</code></pre>
<p>The <code>#!</code> is a “shebang”: a special comment line that tells computer how to run our script</p>
</section>
<section id="now-make-the-script-executable" class="level1">
<h1>Now make the script executable!</h1>
<p>At the terminal:</p>
<pre><code>chmod +x build.sh

# Note &quot;x&quot; when we use `ls -l`
ls -l build.sh

file build.sh

# Now it&#39;s a command/program!
./build.sh</code></pre>
</section>
<section id="taking-a-step-back-to-optimize-our-environment-with-dotfiles" class="level1">
<h1>Taking a step back to optimize our environment with dotfiles</h1>
<p>We’ve already seen dotfiles in use with Tmux. Many other Unix programs accept configuration from dotfiles, including <code>nano</code> and our Unix shell itself.</p>
</section>
<section id="first-some-things-to-make-nano-nicer" class="level1">
<h1>First some things to make nano nicer</h1>
<p>Run <code>nano ~/.nanorc</code>, then enter:</p>
<pre><code>set tabsize 4
set tabstospaces
set autoindent</code></pre>
<p>It’s not perfect, but it’s a start.</p>
</section>
<section id="now-our-environment" class="level1">
<h1>Now our environment</h1>
<p>Tired of running <code>module load intro-bio</code>? Or editing your <code>PATH</code> variable?</p>
<p>Run <code>nano ~/.bashrc</code>:</p>
<pre><code>module load intro-bio
export PATH=~/bin/:$PATH</code></pre>
<p>Now run <code>source ~/.bashrc</code> to reload.</p>
</section>
<section id="back-to-biology" class="level1">
<h1>Back to biology</h1>
</section>
<section id="counting-sequences-per-species" class="level1">
<h1>Counting sequences per species</h1>
<p>In our <code>build.sh</code> file:</p>
<pre><code># Compute number of sequences per species
csvuniq -zc species data/sfv.csv &gt; output/seqs_per_species.csv</code></pre>
<p>Type <code>Ctrl-o</code> to save without exiting.</p>
</section>
<section id="run-and-investigate" class="level1">
<h1>Run and investigate</h1>
<p>In the shell:</p>
<pre><code>ls output

./build.sh

ls output

csvlook output/seqs_per_species.csv</code></pre>
</section>
<section id="add-some-other-steps" class="level1">
<h1>Add some other steps</h1>
<p>In <code>nano build.sh</code> file, add:</p>
<pre><code># Compute number of sequences per specimen
csvuniq -zc specimen,species,location data/sfv.csv &gt; output/seqs_per_specimen.csv

# Use those results to count specimens per species
csvuniq -zc species output/seqs_per_specimen.csv &gt; output/specs_per_species.csv

# Also use them to count specimens by species and location
csvuniq -zc species,location output/seqs_per_specimen.csv &gt; output/specs_per_species_location.csv</code></pre>
</section>
<section id="run-and-investigate-1" class="level1">
<h1>Run and investigate</h1>
<p>In the shell:</p>
<pre><code>./build.sh

csvlook output/seqs_per_specimen.csv | less -S</code></pre>
</section>
<section id="how-shell-variables-work" class="level1">
<h1>How shell variables work</h1>
<p>To clarify and simplify, we introduce <em>variables</em>!</p>
<p>In the shell:</p>
<pre><code># Assign the value &quot;output&quot; to the variable outdir
# Note no spaces around the &quot;=&quot;
outdir=&quot;output&quot;

# We can use that variable in further commands by prepending $
echo $outdir
ls $outdir</code></pre>
</section>
<section id="cleaning-things-up-with-variables" class="level1">
<h1>Cleaning things up with variables</h1>
<p>In our <code>build.sh</code> file:</p>
<pre><code># Specify output directory, and name input data
outdir=&quot;output&quot;
metadata=&quot;data/sfv.csv&quot;


# Compute number of sequences per species
seqs_per_species=&quot;$outdir/seqs_per_species.csv&quot;
csvuniq -zc species $metadata &gt; $seqs_per_species

# Compute number of sequences per specimen
seqs_per_specimen=&quot;$outdir/seqs_per_specimen.csv&quot;
csvuniq -zc specimen,species,location $metadata &gt; $seqs_per_specimen</code></pre>
</section>
<section id="continuing-variable-clean-up" class="level1">
<h1>Continuing variable clean up</h1>
<section id="using-output-variables-as-input" class="level2">
<h2>Using output variables as input:</h2>
<pre><code># Use those results to compute number of specimens per species
specs_per_species=&quot;$outdir/specs_per_species.csv&quot;
csvuniq -zc species $seqs_per_specimen &gt; $specs_per_species

# Also use them to compute number of specimens per species and location
specs_per_species_location=&quot;$outdir/specs_per_species_location.csv&quot;
csvuniq -zc species,location $seqs_per_specimen &gt; $specs_per_species_location</code></pre>
<!-- 18 minutes so far -->
</section>
</section>
<section id="verify-that-it-runs" class="level1">
<h1>Verify that it runs</h1>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">./build.sh</span>

<span class="kw">ls</span> output

<span class="kw">csvlook</span> output/seqs_per_specimen.csv</code></pre>
</section>
<section id="writing-your-own-commands" class="level1">
<h1>Writing your own commands</h1>
<section id="the-other-side-of-shell-scripting-and-fulfillment-of-the-dream-of-composition" class="level2">
<h2>The other side of shell scripting, and fulfillment of the dream of composition</h2>
</section>
</section>
<section id="looking-for-common-command-patterns" class="level1">
<h1>Looking for common command patterns</h1>
<section id="letting-laziness-lead-the-way" class="level2">
<h2>Letting laziness lead the way…</h2>
<p>We’ve seen <code>csvlook ___ | less -S</code> quite a few times now for looking at csv data. So let’s give this a name!</p>
</section>
</section>
<section id="writing-a-csvless-script" class="level1">
<h1>Writing a <code>csvless</code> script</h1>
<p>Open up a new file in <code>~/bin/csvless</code>, and write:</p>
<pre><code>#!/bin/bash

csvlook $@ | less -S</code></pre>
<p>Here <code>$@</code> is a “magic” variable that points to all of the arguments passed to the command.</p>
</section>
<section id="testing-our-csvless-script" class="level1">
<h1>Testing our <code>csvless</code> script</h1>
<p>In the shell:</p>
<pre><code>chmod +x ~/bin/csvless

csvless data/sfv.csv

csvcut -c sequence,specimen,species data/sfv.csv | csvless</code></pre>
</section>
<section id="another-quick-example-csvhead" class="level1">
<h1>Another quick example: <code>csvhead</code></h1>
<p>Open up a new file in <code>~/bin/csvhead</code>, and write:</p>
<pre><code>#!/bin/bash

csvlook $@ | head -n 20</code></pre>
<p>Again, <code>chmod +x</code>, then test it!</p>
</section>
<section id="other-things-to-know-for-shell-scripting" class="level1">
<h1>Other things to know for shell scripting:</h1>
<ul>
<li>Iteration and arrays: Let us nest or loop over things like different species or locations and process separately</li>
<li>Conditionals: Let you run tests to determine what code to run</li>
<li>find/xargs/parallel: Run a command or program in parallel over a sequence of files</li>
</ul>
<p>Most of this we’ll leave to the book…</p>
</section>
<section id="simple-iteration" class="level1">
<h1>Simple iteration</h1>
<p>One line in the terminal:</p>
<pre><code>for i in $(ls); do echo &quot;My file $i is cool&quot;; done</code></pre>
<p>In a file, you can expand this:</p>
<pre><code>for i in $(ls)
do
  echo &quot;My file $i is cool&quot;
done</code></pre>
</section>
<section id="how-we-might-use-this" class="level1">
<h1>How we might use this?</h1>
<p>We could do a series of computations for partitions of our data. For instance:</p>
<pre><code># For each location...
locations=($(csvuniq -c location $metadata | tail -n +2))
for location in ${locations[*]}
do
  # Create a location outdir, if it doesn&#39;t already exist
  loc_outdir=&quot;$outdir/$location&quot;
  mkdir -p $loc_outdir
  
  # Do some computations for $location
done</code></pre>
</section>
<section id="conditionals" class="level1">
<h1>Conditionals</h1>
<pre><code>#!/bin/bash

if [commands]
then
  [if-statements]
else
  [else-statements]
fi</code></pre>
</section>
<section id="conditional-example" class="level1">
<h1>Conditional example</h1>
<pre><code>#!/bin/bash

if grep &quot;pattern&quot; some_file.txt &gt; /dev/null
then
  # commands to run if &quot;pattern&quot; is found
  echo &quot;found &#39;pattern&#39; in &#39;some_file.txt&quot;
else
  echo &quot;no such pattern found...&quot;
fi</code></pre>
<p>Question: Why do we use <code>&gt; /dev/null</code> here?</p>
</section>
<section id="findxargsparallel" class="level1">
<h1>find/xargs/parallel</h1>
<p>There is lots of explanation in the book, so you can learn more there. But to get a brief sense:</p>
<ul>
<li><code>find</code>: Let’s you compute very specific lists of files, returned to <code>stdout</code> separated by lines</li>
<li><code>xargs</code>: Can take line separated items, and apply them as arguments to some command
<ul>
<li>Good for large collections</li>
<li>Can run in parallel</li>
</ul></li>
<li><code>parallel</code>: More robust version of xargs, with better control over parallelization</li>
</ul>
</section>
<section id="build-tools-such-as-scons-and-make-offer-the-following-advantages-over-shell-scripts" class="level1">
<h1>Build tools such as <code>Scons</code> and <code>make</code> offer the following advantages over shell scripts:</h1>
<ul>
<li>Only rebuild what’s needed (very important with long running programs) by tracking:
<ul>
<li>whether data has changed</li>
<li>whether commands have changed</li>
</ul></li>
<li>Generally more robust</li>
<li>Automatically parallelize (sanely)</li>
</ul>
</section>
<section id="excercise-1" class="level1">
<h1>Excercise 1</h1>
<p>Add the alignment and tree building we did in the last two sessions to our <code>build.sh</code> script.</p>
</section>
<section id="excercise-2" class="level1">
<h1>Excercise 2</h1>
<p>Add the following to the end of our <code>build.sh</code> script, and complete the TODO items in comments:</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="co"># For each location...</span>
<span class="ot">locations=($(</span><span class="kw">csvuniq</span> -c location <span class="ot">$metadata</span> <span class="kw">|</span> <span class="kw">tail</span> -n +2<span class="ot">))</span>
<span class="kw">for</span> <span class="kw">location</span> in <span class="ot">${locations[*]}</span>
<span class="kw">do</span>
  <span class="co"># Create a location outdir, if it doesn&#39;t already exist</span>
  <span class="ot">loc_outdir=</span><span class="st">&quot;</span><span class="ot">$outdir</span><span class="st">/</span><span class="ot">$location</span><span class="st">&quot;</span>
  <span class="kw">mkdir</span> -p <span class="ot">$loc_outdir</span>

  <span class="co"># Create a subset of the metadata for just that location</span>
  <span class="ot">loc_metadata=</span><span class="st">&quot;</span><span class="ot">$loc_outdir</span><span class="st">/metadata.csv&quot;</span>
  <span class="kw">csvgrep</span> -c location -m <span class="ot">$location</span> <span class="ot">$metadata</span> <span class="kw">&gt;</span> <span class="ot">$loc_metadata</span>

  <span class="co"># Create a list of sequences sampled from that location</span>
  <span class="ot">loc_sequences=</span><span class="st">&quot;</span><span class="ot">$loc_outdir</span><span class="st">/sequences&quot;</span>
  <span class="kw">csvcut</span> -c sequence <span class="ot">$loc_metadata</span> <span class="kw">&gt;</span> <span class="ot">$loc_sequences</span>

  <span class="co"># TODO:</span>
  <span class="co"># * Subset Ex. 1 alignment to sequences from $location using `seqmagick convert`</span>
  <span class="co"># * Build a phylogenetic tree from this alignment subset using FastTree</span>
<span class="kw">done</span></code></pre>
</section>
<section id="other-exercises" class="level1">
<h1>Other exercises</h1>
<ul>
<li>Write a handy little shell script for doing something (for example, printing the last 20 commands entered)</li>
</ul>
</section>
<section id="reading" class="level1">
<h1>Reading</h1>
<p>Recommended reading:</p>
<ul>
<li>Chapter 12</li>
</ul>
<p>Reading for next class (if you want to read ahead):</p>
<ul>
<li>Chapter 5</li>
</ul>
</section>
<section id="resources" class="level1">
<h1>Resources</h1>
<ul>
<li>SCons (sane <code>make</code> alternative) for bioinformatics post - <a href="http://www.metasoarous.com/scons-for-data-science-and-compbio/" class="uri">http://www.metasoarous.com/scons-for-data-science-and-compbio/</a></li>
<li>seqmagick - <a href="http://fhcrc.github.io/seqmagick/" class="uri">http://fhcrc.github.io/seqmagick/</a> for munging sequence data</li>
<li><code>nano</code> for simple file editing on the rhinos</li>
<li>Vim for more powerful text editing - (type <code>vimtutor</code> at the terminal…)</li>
<li>Erick’s favorite <a href="https://www.youtube.com/watch?v=olH-9b3VJfs">video</a> about shell scripting, and <a href="http://shellhaters.org/">website</a></li>
</ul>
</section>
<section id="spoiler-alert" class="level1">
<h1>SPOILER ALERT!!!</h1>
<section id="the-following-slides-contain-solutions-to-the-exercises." class="level2">
<h2>The following slides contain solutions to the exercises.</h2>
<p>Go no further if you’re keen to solve the problems yourself.</p>
</section>
</section>
<section id="ex.-1-automating-alignment-and-tree-building" class="level1">
<h1>Ex. 1: Automating alignment and tree building</h1>
<pre><code>#!/bin/bash
# ...

inseqs=&quot;data/sfv.fasta&quot;

# Alignment
alignment=&quot;$outdir/alignment.fasta&quot;
muscle -maxiters 2 -in $inseqs -out $alignment

# Tree
tree=&quot;$outdir/tree.nw&quot;
FastTree -nt $alignment &gt; $tree</code></pre>
</section>
<section id="ex.-2-finish-location-for-loop" class="level1">
<h1>Ex. 2: Finish location for loop</h1>
<pre><code>  # Subset our alignment to just that location
  loc_alignment=&quot;$loc_outdir/alignment.fasta&quot;
  seqmagick convert --include-from-file $loc_sequences $alignment $loc_alignment

  # Build a location tree
  loc_tree=&quot;$loc_outdir/tree.nw&quot;
  FastTree -nt $loc_alignment</code></pre>
</section>
<section id="this-slide-intentionally-left-almost-blank" class="level1">
<h1>This slide intentionally left almost blank…</h1>
<p><a href="http://fredhutchio.github.io/intro-bioinformatics">Back to homepage</a></p>
</section>
</div>

<script src="reveal.js/lib/js/head.min.js"></script>
<script src="reveal.js/js/reveal.min.js"></script>

<script>
  // Full list of configuration options available here:
  // https://github.com/hakimel/reveal.js#configuration
  Reveal.initialize({
    controls: true,
    progress: true,
    history: true,
    center: false,


    rollingLinks: false,
  
  // available themes are in /css/theme
      theme: Reveal.getQueryHash().theme || 'csmall',
  
  // default/cube/page/concave/zoom/linear/fade/none
      transition: Reveal.getQueryHash().transition || 'fade',
  
  // Optional libraries used to extend on reveal.js
  dependencies: [
    { src: 'reveal.js/lib/js/classList.js', condition: function() { return !document.body.classList; } },
    { src: 'reveal.js/plugin/markdown/showdown.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
    { src: 'reveal.js/plugin/markdown/markdown.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
    { src: 'reveal.js/plugin/highlight/highlight.js', async: true, callback: function() { hljs.initHighlightingOnLoad(); } },
    { src: 'reveal.js/plugin/zoom-js/zoom.js', async: true, condition: function() { return !!document.body.classList; } },
    { src: 'MathJax/MathJax.js?config=TeX-AMS-MML_HTMLorMML'}
    // { src: 'reveal.js/plugin/notes/notes.js', async: true, condition: function() { return !!document.body.classList; } }
    // { src: 'reveal.js/plugin/remotes/remotes.js', async: true, condition: function() { return !!document.body.classList; } }
  ]
  });
Reveal.addEventListener( 'slidechanged', function( event ) {
    MathJax.Hub.Rerender();
} );
</script>

</body>
</html>
