<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">

<title>Lesson 3 - Unix Scripting &amp; Automation</title>

<meta name="description" content="Lesson 3 - Unix Scripting &amp; Automation">

<!-- Disabled for now - embeds HTML in author string -->
<!--  <meta name="author" content="Introduction to Bioinformatics" /> -->

<meta name="apple-mobile-web-app-capable" content="yes" />
<meta name="apple-mobile-web-app-status-bar-style" content="black-translucent" />

<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">

<link rel="stylesheet" href="reveal.js/css/reveal.css">
  <link rel="stylesheet" href="src/theme/csmall.css" id="theme">


<!-- For syntax highlighting -->
  <link rel="stylesheet" href="reveal.js/lib/css/zenburn.css">


<!-- If the query includes 'print-pdf', use the PDF print sheet -->
<script>
  document.write( '<link rel="stylesheet" href="reveal.js/css/print/' +
    ( window.location.search.match( /print-pdf/gi ) ? 'pdf' : 'paper' ) +
    '.css" type="text/css" media="print">' );
</script>

<!--[if lt IE 9]>
<script src="reveal.js/lib/js/html5shiv.js"></script>
<![endif]-->

<!-- Note escaped dollars -->
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      tex2jax: {inlineMath: [['$','$'], ['\\(','\\)']]}
    });
  </script>

</head>

<body>

<div class="reveal">

<!-- Any section element inside of this container is displayed as a slide -->
<div class="slides">

<section>
<h1>Lesson 3 - Unix Scripting &amp; Automation</h1>
<h3>Introduction to Bioinformatics</h3>
<p>
<h4></h4>
</p>
</section>


<section id="first-things-first" class="level1">
<h1>First things first</h1>
<ul>
<li>Were you able to get <code>csvuniq</code> installed/working (check <code>~/bin/csvuniq -h</code>)?</li>
<li>How is project planning and outside time going?</li>
</ul>
</section>
<section id="shell-scripting" class="level1">
<h1>Shell scripting</h1>
<p><em>In which our flying, speaking beasts assemble and become legion</em></p>
</section>
<section id="a-shell-script-is-just-a-list-of-commands-to-be-executed" class="level1">
<h1>A <em>shell script</em> is just a list of commands to be executed</h1>
<p>We can type these commands into text files, and instruct our Unix shell to execute them.</p>
</section>
<section id="surprise" class="level1">
<h1>Surprise!</h1>
<p>The Unix shell is also a mini-programming language!</p>
<p>Next to piping, this is our second route towards the promise of Unix composability</p>
</section>
<section id="for-bioinformatics-this-means" class="level1">
<h1>For bioinformatics this means</h1>
<ul>
<li>Simple analysis tools/commands</li>
<li>Automation</li>
</ul>
</section>
<section id="shell-scripting-is-just-duct-tape" class="level1">
<h1>Shell scripting is just “duct tape”</h1>
<ul>
<li>Perfect for connecting existing shell commands together</li>
<li>Less so for more complicated data flow and logic (for which we have python, R, etc.)</li>
</ul>
</section>
<section id="our-environment" class="level1">
<h1>Our environment</h1>
<p>We can also script our environment via a dotfile that gets loaded when shell sessions are created: <code>~/.bashrc</code>.</p>
</section>
<section id="our-environment-1" class="level1">
<h1>Our environment</h1>
<ul>
<li>Run <code>vim ~/.bashrc</code> to open the file</li>
<li>hit <code>i</code> for insert mode</li>
<li>use arrow keys to scroll to the bottom of the file, then add:</li>
</ul>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">module</span> load intro-bio
<span class="ex">module</span> load R/3.3.3-foss-2016b-fh2 <span class="co"># for now; we&#39;ll put this in intro-bio soon</span>
<span class="bu">export</span> <span class="va">PATH=</span>~/bin/:<span class="va">$PATH</span></code></pre></div>
<p>Save and quit (<code>Esc</code> then <code>:wq&lt;Enter&gt;</code>), then run <code>source ~/.bashrc</code> to reload the changes (for each already open tmux pane).</p>
</section>
<section id="back-to-biology" class="level1">
<h1>Back to biology</h1>
</section>
<section id="our-first-script" class="level1">
<h1>Our first script!</h1>
<p>We’re going to create a <code>build.sh</code> script to automate the analyses we did last week.</p>
</section>
<section id="our-first-script-1" class="level1">
<h1>Our first script!</h1>
<p>Run <code>vim build.sh</code> to open the file, then hit <code>i</code> for insert mode, and enter the following text:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co"># Compute number of sequences per species</span>
<span class="ex">csvuniq</span> -zc species data/sfv.csv <span class="op">&gt;</span> output/seqs_per_species.csv
<span class="bu">echo</span> <span class="st">&quot;IT WORKED!&quot;</span></code></pre></div>
<ul>
<li>Hit <code>Esc</code> to return to command mode</li>
<li>Use the command <code>:w&lt;Enter&gt;</code> to save the file</li>
</ul>
</section>
<section id="running-our-script" class="level1">
<h1>Running our script</h1>
<ul>
<li>Open a new tmux pane with <code>Ctrl-a |</code> (or switch to an existing one with <code>Ctrl-a &lt;arrow-key&gt;</code>)</li>
<li>Then run the following in the shell:</li>
</ul>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="bu">cd</span> ~/bioinfclass
<span class="fu">bash</span> build.sh

<span class="fu">ls</span> output <span class="co"># note the new file</span>

<span class="ex">csvlook</span> output/seqs_per_species.csv <span class="co"># take a look</span></code></pre></div>
</section>
<section id="making-into-a-proper-command" class="level1">
<h1>Making into a proper command</h1>
<p>Go back to your <code>vim build.sh</code> tmux pane (<code>Ctrl-a &lt;arrow-key&gt;</code>), and add the following lines:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">#!/bin/bash</span>
<span class="co"># ^ this magic comment is a &quot;shebang&quot;; It tells the shell how to interpret our script.</span>

<span class="co"># Sane error handling settings (stop running for most errors)</span>
<span class="kw">set</span> <span class="ex">-euf</span> -o pipefail

<span class="co"># Compute number of sequences per species</span>
<span class="ex">csvuniq</span> -zc species data/sfv.csv <span class="op">&gt;</span> output/seqs_per_species.csv
<span class="bu">echo</span> <span class="st">&quot;IT WORKED!&quot;</span></code></pre></div>
<p>Save when done (<code>Esc</code> for command mode, then <code>:w&lt;Enter&gt;</code>).</p>
</section>
<section id="now-make-the-script-executable" class="level1">
<h1>Now make the script executable!</h1>
<p>Switch back to your unix shell tmux pane (<code>Ctrl-a &lt;arrow-key&gt;</code>), and then execute:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">chmod</span> +x build.sh

<span class="co"># Note &quot;x&quot; when we use `ls -l`</span>
<span class="fu">ls</span> -l build.sh

<span class="fu">file</span> build.sh

<span class="co"># Now it&#39;s a command/program!</span>
<span class="ex">./build.sh</span></code></pre></div>
</section>
<section id="back-to-our-build-script" class="level1">
<h1>Back to our build script</h1>
<p>Go back to your <code>vim build.sh</code> tmux pane, (<code>Ctrl-a &lt;arrow-key&gt;</code>), press down till you get to the bottom of the file, and add to the end:</p>
<pre><code># Compute number of sequences per specimen
csvuniq -zc specimen,species,location data/sfv.csv &gt; output/seqs_per_specimen.csv

# Use those results to count specimens per species
csvuniq -zc species output/seqs_per_specimen.csv &gt; output/specs_per_species.csv

# Also use them to count specimens by species and location
csvuniq -zc species,location output/seqs_per_specimen.csv &gt; output/specs_per_species_location.csv</code></pre>
</section>
<section id="run-and-investigate" class="level1">
<h1>Run and investigate</h1>
<p>In the shell:</p>
<pre><code>./build.sh

ls output

csvlook output/seqs_per_specimen.csv | less -S</code></pre>
</section>
<section id="questions" class="level1">
<h1>Questions?</h1>
</section>
<section id="shell-variables" class="level1">
<h1>Shell variables</h1>
<p>Shell variables act as shorthand references to paths, files, or data. What the variable name points to can change, without having to change (much) of the code. This makes it easier to grow your scripts.</p>
</section>
<section id="shell-variables-1" class="level1">
<h1>Shell variables</h1>
<p>In the shell:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co"># Set the shell variable `birds`</span>
<span class="co"># (Note: no spaces around the &quot;=&quot;)</span>
<span class="va">birds=</span><span class="st">&quot;are really dinosaurs&quot;</span>

<span class="co"># We can use that variable in further commands by prepending $</span>
<span class="bu">echo</span> <span class="va">$birds</span></code></pre></div>
</section>
<section id="cleaning-up-our-build-script" class="level1">
<h1>Cleaning up our build script</h1>
<p>Edit our <code>build.sh</code> file to look as follows:</p>
<pre><code>#!/bin/bash
set -euf -o pipefail

# Specify output directory, and name input data
outdir=&quot;output&quot;
metadata=&quot;data/sfv.csv&quot;

# Compute number of sequences per species
seqs_per_species=&quot;$outdir/seqs_per_species.csv&quot;
csvuniq -zc species $metadata &gt; $seqs_per_species

# Compute number of sequences per specimen
seqs_per_specimen=&quot;$outdir/seqs_per_specimen.csv&quot;
csvuniq -zc specimen,species,location $metadata &gt; $seqs_per_specimen</code></pre>
</section>
<section id="cleaning-up-our-build-script-1" class="level1">
<h1>Cleaning up our build script</h1>
<p>Now note that when we add the following, we get to refer to past results using the defined variables:</p>
<pre><code># Use seqs_per_specimen to compute specimens per species
specs_per_species=&quot;$outdir/specs_per_species.csv&quot;
csvuniq -zc species $seqs_per_specimen &gt; $specs_per_species

# Also use them to compute number of specimens per species and location
specs_per_species_location=&quot;$outdir/specs_per_species_location.csv&quot;
csvuniq -zc species,location $seqs_per_specimen &gt; $specs_per_species_location</code></pre>
<p>Save when done (<code>Esc</code> then <code>:w&lt;Enter&gt;</code>).</p>
</section>
<section id="verify-that-it-runs" class="level1">
<h1>Verify that it runs</h1>
<p>Back in your unix shell (<code>Ctrl-a &lt;arrow-key&gt;</code>):</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">./build.sh</span>

<span class="fu">ls</span> output

<span class="ex">csvlook</span> output/seqs_per_specimen.csv</code></pre></div>
</section>
<section id="questions-1" class="level1">
<h1>Questions</h1>
</section>
<section id="writing-your-own-commands" class="level1">
<h1>Writing your own commands</h1>
<section id="the-other-side-of-shell-scripting-and-fulfillment-of-the-dream-of-composition" class="level2">
<h2>The other side of shell scripting, and fulfillment of the dream of composition</h2>
</section>
</section>
<section id="looking-for-common-command-patterns" class="level1">
<h1>Looking for common command patterns</h1>
<section id="letting-laziness-lead-the-way" class="level2">
<h2>Letting laziness lead the way…</h2>
<p>We’ve seen <code>csvlook ___ | less -S</code> quite a few times now for looking at csv data. So let’s give this a name!</p>
</section>
</section>
<section id="writing-a-csvless-script" class="level1">
<h1>Writing a <code>csvless</code> script</h1>
<p>Open up a new file with <code>vim ~/bin/csvless</code>, and insert:</p>
<pre><code>#!/bin/bash

csvlook $@ | less -S</code></pre>
<p>Here <code>$@</code> is a “magic” variable that points to all of the arguments passed to the command. If you didn’t want positional arguments, you can use <code>$1</code> or <code>$2</code> for the 1st or 2nd (etc.) positional arguments passed to the script.</p>
</section>
<section id="testing-our-csvless-script" class="level1">
<h1>Testing our <code>csvless</code> script</h1>
<p>Save and close the file (<code>Esc</code> then <code>:wq&lt;Enter&gt;</code>), then run the following from the shell:</p>
<pre><code>chmod +x ~/bin/csvless

csvless data/sfv.csv

csvcut -c sequence,specimen,species data/sfv.csv | csvless</code></pre>
</section>
<section id="another-quick-example-csvhead" class="level1">
<h1>Another quick example: <code>csvhead</code></h1>
<p>Open up a new file with <code>vim ~/bin/csvhead</code>, and insert:</p>
<pre><code>#!/bin/bash

csvlook $@ | head -n 20</code></pre>
<p>Again, save and close (<code>Esc</code> + <code>:wq&lt;Enter&gt;</code>), run <code>chmod +x ~/bin/csvhead</code>, then test it!</p>
</section>
<section id="other-things-to-know-for-shell-scripting-power-usage" class="level1">
<h1>Other things to know for shell scripting &amp; power usage:</h1>
<ul>
<li>Iteration and arrays: Let us nest or loop over things like different species or locations and process separately</li>
<li>Conditionals: Let you run tests to determine what code to run</li>
<li>find/xargs/parallel: Run a command or program in parallel over a sequence of files</li>
</ul>
<p>Most of this we’ll leave to the book…</p>
</section>
<section id="simple-iteration" class="level1">
<h1>Simple iteration</h1>
<p>One line in the terminal:</p>
<pre><code>for i in $(ls); do echo &quot;My file $i is cool&quot;; done</code></pre>
<p>In a file, you can expand this:</p>
<pre><code>for i in $(ls)
do
  echo &quot;My file $i is cool&quot;
done</code></pre>
</section>
<section id="how-we-might-use-this" class="level1">
<h1>How we might use this?</h1>
<p>We could do a series of computations for partitions of our data. For instance:</p>
<pre><code># For each location...
locations=($(csvuniq -c location $metadata | tail -n +2))
for location in ${locations[*]}
do
  # Create a location outdir, if it doesn&#39;t already exist
  loc_outdir=&quot;$outdir/$location&quot;
  mkdir -p $loc_outdir
  
  # Do some computations for $location
done</code></pre>
</section>
<section id="conditionals" class="level1">
<h1>Conditionals</h1>
<pre><code>#!/bin/bash

if [commands]
then
  [if-statements]
else
  [else-statements]
fi</code></pre>
</section>
<section id="conditional-example" class="level1">
<h1>Conditional example</h1>
<pre><code>#!/bin/bash

if grep &quot;pattern&quot; some_file.txt &gt; /dev/null
then
  # commands to run if &quot;pattern&quot; is found
  echo &quot;found &#39;pattern&#39; in &#39;some_file.txt&quot;
else
  echo &quot;no such pattern found...&quot;
fi</code></pre>
<p>Question: Why do we use <code>&gt; /dev/null</code> here?</p>
</section>
<section id="findxargsparallel" class="level1">
<h1>find/xargs/parallel</h1>
<p>There is lots of explanation in the book, so you can learn more there. But to get a brief sense:</p>
<ul>
<li><code>find</code>: Lets you compute very specific lists of files, returned to <code>stdout</code> separated by lines</li>
<li><code>xargs</code>: Can take line separated items, and apply them as arguments to some command
<ul>
<li>Good for large collections</li>
<li>Can run in parallel</li>
</ul></li>
<li><code>parallel</code>: More robust version of xargs, with better control over parallelization</li>
</ul>
</section>
<section id="build-tools-such-as-scons-and-make-offer-the-following-advantages-over-shell-scripts" class="level1">
<h1>Build tools such as <code>Scons</code> and <code>make</code> offer the following advantages over shell scripts:</h1>
<ul>
<li>Only rebuild what’s needed (very important with long running programs) by tracking:
<ul>
<li>whether data has changed</li>
<li>whether commands have changed</li>
</ul></li>
<li>Generally more robust</li>
<li>Automatically parallelize (sanely)</li>
</ul>
</section>
<section id="excercise-1" class="level1">
<h1>Excercise 1</h1>
<section id="section" class="level2">
<h2></h2>
<p>How would we add the alignment and tree building we did in the last two sessions to our <code>build.sh</code> script?</p>
</section>
<section id="section-1" class="level2">
<h2></h2>
<p>Make the top of our script look like:</p>
<pre><code>#!/bin/bash
# ...

inseqs=&quot;data/sfv.fasta&quot;

# Alignment
alignment=&quot;$outdir/alignment.fasta&quot;
muscle -maxiters 2 -in $inseqs -out $alignment

# Tree
tree=&quot;$outdir/tree.nw&quot;
FastTree -nt $alignment &gt; $tree</code></pre>
</section>
</section>
<section id="excercise-2" class="level1">
<h1>Excercise 2</h1>
<section id="say-we-want-to-do-some-analyses-for-each-location." class="level2">
<h2>Say we want to do some analyses for each location.</h2>
<p>Let’s start by creating a separate directory for each. How would we add location specific metadata using our csvkit tools?</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co"># For each location...</span>
<span class="va">locations=($(</span><span class="ex">csvuniq</span> -c location <span class="va">$metadata</span> <span class="kw">|</span> <span class="fu">tail</span> -n +2<span class="va">))</span>
<span class="kw">for</span> <span class="ex">location</span> in <span class="va">${locations[*]}</span>
<span class="kw">do</span>
  <span class="co"># Create a location outdir, if it doesn&#39;t already exist</span>
  <span class="va">loc_outdir=</span><span class="st">&quot;</span><span class="va">$outdir</span><span class="st">/</span><span class="va">$location</span><span class="st">&quot;</span>
  <span class="fu">mkdir</span> -p <span class="va">$loc_outdir</span>

  <span class="co"># Create a subset of the metadata for just that location</span>
  <span class="va">loc_metadata=</span><span class="st">&quot;</span><span class="va">$loc_outdir</span><span class="st">/metadata.csv&quot;</span>
  <span class="co"># XXX TODO! Add code here to create $loc_metadata</span>

<span class="kw">done</span></code></pre></div>
</section>
<section id="section-2" class="level2">
<h2></h2>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co"># For each location...</span>
<span class="va">locations=($(</span><span class="ex">csvuniq</span> -c location <span class="va">$metadata</span> <span class="kw">|</span> <span class="fu">tail</span> -n +2<span class="va">))</span>
<span class="kw">for</span> <span class="ex">location</span> in <span class="va">${locations[*]}</span>
<span class="kw">do</span>
  <span class="co"># Create a location outdir, if it doesn&#39;t already exist</span>
  <span class="va">loc_outdir=</span><span class="st">&quot;</span><span class="va">$outdir</span><span class="st">/</span><span class="va">$location</span><span class="st">&quot;</span>
  <span class="fu">mkdir</span> -p <span class="va">$loc_outdir</span>

  <span class="co"># Create a subset of the metadata for just that location</span>
  <span class="va">loc_metadata=</span><span class="st">&quot;</span><span class="va">$loc_outdir</span><span class="st">/metadata.csv&quot;</span>
  <span class="ex">csvgrep</span> -c location -m <span class="va">$location</span> <span class="va">$metadata</span> <span class="op">&gt;</span> <span class="va">$loc_metadata</span>

<span class="kw">done</span></code></pre></div>
</section>
</section>
<section id="excercise-3" class="level1">
<h1>Excercise 3</h1>
<section id="how-would-we-create-a-phylogenetic-tree-for-each-location-in-our-data-set" class="level2">
<h2>How would we create a phylogenetic tree for each location in our data set?</h2>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co"># For each location...</span>
<span class="va">locations=($(</span><span class="ex">csvuniq</span> -c location <span class="va">$metadata</span> <span class="kw">|</span> <span class="fu">tail</span> -n +2<span class="va">))</span>
<span class="kw">for</span> <span class="ex">location</span> in <span class="va">${locations[*]}</span>
<span class="kw">do</span>
  <span class="co"># ... create location metadata, etc.</span>

  <span class="co"># Create a list of sequences sampled from that location</span>
  <span class="va">loc_sequences=</span><span class="st">&quot;</span><span class="va">$loc_outdir</span><span class="st">/sequences&quot;</span>
  <span class="ex">csvcut</span> -c sequence <span class="va">$loc_metadata</span> <span class="op">&gt;</span> <span class="va">$loc_sequences</span>

  <span class="co"># TODO:</span>
  <span class="co"># * Subset Ex.1 alignment using `seqmagick convert`</span>
  <span class="co"># * Build a tree from this alignment subset using FastTree</span>
<span class="kw">done</span></code></pre></div>
</section>
<section id="section-3" class="level2">
<h2></h2>
<p>Add the following to the end of our for loop</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co"># For each location...</span>
<span class="va">locations=($(</span><span class="ex">csvuniq</span> -c location <span class="va">$metadata</span> <span class="kw">|</span> <span class="fu">tail</span> -n +2<span class="va">))</span>
<span class="kw">for</span> <span class="ex">location</span> in <span class="va">${locations[*]}</span>
<span class="kw">do</span>
  <span class="co"># ... create location metadata, etc.</span>

  <span class="co"># Create a list of sequences sampled from that location</span>
  <span class="va">loc_sequences=</span><span class="st">&quot;</span><span class="va">$loc_outdir</span><span class="st">/sequences&quot;</span>
  <span class="ex">csvcut</span> -c sequence <span class="va">$loc_metadata</span> <span class="op">&gt;</span> <span class="va">$loc_sequences</span>

  <span class="co"># Subset our alignment to just that location</span>
  <span class="va">loc_alignment=</span><span class="st">&quot;</span><span class="va">$loc_outdir</span><span class="st">/alignment.fasta&quot;</span>
  <span class="ex">seqmagick</span> convert --include-from-file <span class="va">$loc_sequences</span> <span class="va">$alignment</span> <span class="va">$loc_alignment</span>

  <span class="co"># Build a location tree</span>
  <span class="va">loc_tree=</span><span class="st">&quot;</span><span class="va">$loc_outdir</span><span class="st">/tree.nw&quot;</span>
  <span class="ex">FastTree</span> -nt <span class="va">$loc_alignment</span> <span class="op">&gt;</span> <span class="va">$loc_tree</span>

<span class="kw">done</span></code></pre></div>
</section>
</section>
<section id="other-exercises" class="level1">
<h1>Other exercises</h1>
<ul>
<li>Write a handy little shell script for doing something (for example, printing the last 20 commands entered)</li>
</ul>
</section>
<section id="reading" class="level1">
<h1>Reading</h1>
<p>Recommended reading:</p>
<ul>
<li>Chapter 12</li>
</ul>
<p>Reading for next class (if you want to read ahead):</p>
<ul>
<li>Chapter 5</li>
</ul>
</section>
<section id="resources" class="level1">
<h1>Resources</h1>
<ul>
<li>SCons for bioinformatics post - <a href="http://www.metasoarous.com/scons-for-data-science-and-compbio/" class="uri">http://www.metasoarous.com/scons-for-data-science-and-compbio/</a></li>
<li>Seqmagick - <a href="http://fhcrc.github.io/seqmagick/" class="uri">http://fhcrc.github.io/seqmagick/</a> for munging sequence data</li>
<li>Vim for more powerful text editing - (type <code>vimtutor</code> at the terminal…)</li>
<li>Erick’s favorite <a href="https://www.youtube.com/watch?v=olH-9b3VJfs">video</a> about shell scripting, and <a href="http://shellhaters.org/">website</a></li>
</ul>
</section>
<section id="this-slide-intentionally-left-almost-blank" class="level1">
<h1>This slide intentionally left almost blank…</h1>
<p><a href="http://fredhutchio.github.io/intro-bioinformatics">Back to homepage</a></p>
</section>
</div>

<script src="reveal.js/lib/js/head.min.js"></script>
<script src="reveal.js/js/reveal.min.js"></script>

<script>
  // Full list of configuration options available here:
  // https://github.com/hakimel/reveal.js#configuration
  Reveal.initialize({
    controls: true,
    progress: true,
    history: true,
    center: false,
    slideNumber: true,


    rollingLinks: false,
  
  // available themes are in /css/theme
      theme: Reveal.getQueryHash().theme || 'csmall',
  
  // default/cube/page/concave/zoom/linear/fade/none
      transition: Reveal.getQueryHash().transition || 'fade',
  
  // Optional libraries used to extend on reveal.js
  dependencies: [
    { src: 'reveal.js/lib/js/classList.js', condition: function() { return !document.body.classList; } },
    { src: 'reveal.js/plugin/markdown/showdown.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
    { src: 'reveal.js/plugin/markdown/markdown.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
    { src: 'reveal.js/plugin/highlight/highlight.js', async: true, callback: function() { hljs.initHighlightingOnLoad(); } },
    { src: 'reveal.js/plugin/zoom-js/zoom.js', async: true, condition: function() { return !!document.body.classList; } },
    { src: 'MathJax/MathJax.js?config=TeX-AMS-MML_HTMLorMML'}
    // { src: 'reveal.js/plugin/notes/notes.js', async: true, condition: function() { return !!document.body.classList; } }
    // { src: 'reveal.js/plugin/remotes/remotes.js', async: true, condition: function() { return !!document.body.classList; } }
  ]
  });
Reveal.addEventListener( 'slidechanged', function( event ) {
    MathJax.Hub.Rerender();
} );
</script>

</body>
</html>
